crates/chain/src/local_chain.rs

```diff
@@ -1,14 +1,116 @@
+//! The [`LocalChain`] is a local implementation of [`ChainOracle`].
+
 use core::convert::Infallible;
 
-use alloc::collections::{BTreeMap, BTreeSet};
+use crate::collections::BTreeMap;
+use crate::{BlockId, ChainOracle};
+use alloc::{sync::Arc, vec::Vec};
 use bitcoin::BlockHash;
 
-use crate::{BlockId, ChainOracle};
+/// A structure that represents changes to [`LocalChain`].
+pub type ChangeSet = BTreeMap<u32, Option<BlockHash>>;
+
+/// Represents a block of [`LocalChain`].
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct CheckPoint {
+    /// Block height.
+    pub height: u32,
+    /// Block hash.
+    pub hash: BlockHash,
+    /// Previous checkpoint (if any).
+    pub prev: Option<Arc<CheckPoint>>,
+}
+
+impl CheckPoint {
+    /// Get the [`BlockId`] of the checkpoint.
+    pub fn block_id(&self) -> BlockId {
+        BlockId {
+            height: self.height,
+            hash: self.hash,
+        }
+    }
+
+    /// Detach this checkpoint from the previous.
+    pub fn detach(self: Arc<Self>) -> Arc<Self> {
+        Arc::new(Self {
+            height: self.height,
+            hash: self.hash,
+            prev: None,
+        })
+    }
+
+    /// Iterate
+    pub fn iter(self: &Arc<Self>) -> CheckPointIter {
+        CheckPointIter {
+            current: Some(Arc::clone(self)),
+        }
+    }
+}
+
+/// A structure that iterates over checkpoints backwards.
+pub struct CheckPointIter {
+    current: Option<Arc<CheckPoint>>,
+}
+
+impl Iterator for CheckPointIter {
+    type Item = Arc<CheckPoint>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let current = self.current.clone()?;
+        self.current = current.prev.clone();
+        Some(current)
+    }
+}
 
 /// This is a local implementation of [`ChainOracle`].
 #[derive(Debug, Default, Clone, PartialEq, Eq, PartialOrd, Ord)]
 pub struct LocalChain {
-    blocks: BTreeMap<u32, BlockHash>,
+    checkpoints: BTreeMap<u32, Arc<CheckPoint>>,
+}
+
+impl FromIterator<Arc<CheckPoint>> for LocalChain {
```

Comment:
Not sure what this is for. You can create a `LocalChain` from a single `&CheckPoint` by going backwards. Why do you want to create a localchain from an iterator of `CheckPoint`. Also since I think `CheckPoint` should be private I also think this probably shouldn't exist.

Reply:
Removed.